### ISSUE TITLE
weak_computed_property false positive in protocol declaration

### ISSUE THREAD

**Issue by nwhittaker** [realm/SwiftLint#2697]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

A weak protocol property declaration is tripping the weak_computed_property rule.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ echo "protocol { weak var prop: AnyObject { get set } }" | swiftlint lint --no-cache --use-stdin --enable-all-rules

<nopath>:1:12: warning: Weak Computed Property Violation: Adding weak to a computed property has no effect. (weak_computed_property)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.31.0
* Installation method used (Homebrew, CocoaPods, building from source, etc)? Homebrew
* Paste your configuration file: n/a

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
* Which Xcode version are you using (check `xcode-select -p`)? 10.2 (10E125)
* Do you have a sample that shows the issue?

```swift
// This triggers a violation:
protocol {
    weak var prop: AnyObject { get set }
}
```

**Comment by zntfdr** [comment]
The SwiftLint output is correct:
adding weak to a computed property has no effect.

You can see a related stack overflow question [here](https://stackoverflow.com/a/47700104/5840990).

```swift
protocol Test {
    weak var prop: AnyObject? { get set }
}
```

I just tried to run the code above in Xcode 10.1 and even Xcode says so: 
```
'weak' should not be applied to a property declaration in a protocol and will be disallowed in future versions
```

And, to confirm, this is now disallowed in Xcode 10.2:
```
'weak' cannot be applied to a property declaration in a protocol
```
<img width="784" alt="Screen Shot 2019-03-29 at 11 28 36" src="https://user-images.githubusercontent.com/5277837/55209542-d4d48880-5215-11e9-874e-f635a20266f9.png">

**Comment by nwhittaker** [comment]
>The SwiftLint output is correct:
adding weak to a computed property has no effect.

@zntfdr, `weak var prop: AnyObject? { get set }` is not a computed property, though(?). I'd expect to only see the complier warning/error on that line, not also swiftlint's `weak_computed_property` warning.

**Comment by zntfdr** [comment]
My bad, you're right: 
I read it as "declared property", not "computed property" 😅 .

Either way, as mentioned in my previous reply, you shouldn't write `weak` in a protocol property declaration.

With that being said, your point in the issue above is still valid 👍

**Comment by marcelofabri** [comment]
I think the rule behavior is "correct", I just named it poorly. Given that Swift 5 warns on this, I think we can keep it as is (and deprecate/remove this rule) in the future.

Or change the message if it catches a violation inside a protocol declaration. Thoughts?

**Comment by zntfdr** [comment]
I'd completely skip the "violation" if it is caught inside a protocol declaration, as Xcode takes care of this for us already 👍 .

**Comment by marcelofabri** [comment]
> as Xcode takes care of this for us already 👍 .

Only when using Swift 5+, which some teams might take a while to adopt.

**Comment by zntfdr** [comment]
> @marcelofabri: Only when using Swift 5+, which some teams might take a while to adopt.

Even when using Xcode 10.1/Swift 4.2, you get a warning 😃 
```
'weak' should not be applied to a property declaration in a protocol and will be disallowed in future versions
```
Example from [my reply above](https://github.com/realm/SwiftLint/issues/2697#issuecomment-477863126):
<img width="1178" alt="Screen Shot 2019-04-02 at 10 03 52" src="https://user-images.githubusercontent.com/5277837/55373234-b2e94780-552e-11e9-976f-4ef3b2bd44d2.png">

**Comment by jpsim** [comment]
Closing as by design. Let me know if I misunderstood something.

**Comment by maxim4to** [comment]
Once you add Objc compatibility to protocol `weak` is getting allowed. Seems that it is still an issue.
(Swift 5, XCode 10.3)
<img width="1072" alt="Снимок экрана 2019-09-02 в 18 21 50" src="https://user-images.githubusercontent.com/3517394/64124055-ac494f80-cdae-11e9-81c4-5b0ab426ef6d.png">
